### ISSUE TITLE
run non-const Display/Debug impls for const panic message

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#86364]
Given that the CTFE core engine already supports the entire display machinery (demonstrated by the fact that it works in Miri), we might ~~alternatively~~ poke a small hole into the const checks that allows executing formatting machinery from `panic_fmt` only -- something like "miri unleashed", but more targeted.

_Originally posted by @RalfJung in https://github.com/rust-lang/rust/issues/78356#issuecomment-862283228_

Basically the same rationale as allowing -> _ return types in the parser, to get a better error later. In this case, when you use panic!() in CTFE, you are going to fail the build *anyway*, so we can just as well run something that may fail to get evaluated because it does something horribly unconst.

The only problem is that then const fn can execute non-const code at runtime during the panic machinery and before actually panicking, which may not be desirable.

**Comment by RalfJung** [comment]
My thinking was that we'd still have some heavy checks on the code `format_args!` expands to inside `const fn` to ensure that it does not run user-defined `Display`/`Debug` -- basically, we can just have a list of types that are permitted here, and reject the rest.
So, `Arguments::new_v1` would only be allowed under some very special conditions inside `const fn`.

(This can theoretically be side-stepped with `transmute`, but whoever does that really can't expect *any* kind of guarantees^^.)

**Comment by Aaron1011** [comment]
> In this case, when you use panic!() in CTFE, you are going to fail the build anyway, so we can just as well run something that may fail to get evaluated because it does something horribly unconst.

This is only true because we don't currently support `const catch_unwind` (and const unwinding is only implemented in Miri, not the CTFE). If we ever decided to support `const catch_unwind`, then we would need to make catching an 'unconst panic' into a hard error.

**Comment by oli-obk** [comment]
True, but I personally believe we should never support that. It's not really something that is useful in CTFE. Catching panics is solely there to prevent running across FFI boundaries or to create somewhat isolated units like threads or e.g. tabs or tasks in applications. CTFE is already perfectly isolated, and since there is no unwinding, there's no problem with FFI boundaries either. There's also no resources that need to be freed, as the entire evaluator is freed after every (successful or not) evaluation, including the entire virtual memory (modulo some things like vtables).

**Comment by oli-obk** [comment]
Almost finished impl that can probably be revived:  #90488

**Comment by RalfJung** [comment]
Note however that it will need some adjustments because we don't want to run any user-defined formatting code, only the code in std.
